File: leetcode/BackTracking/leetcode-78/78.go
package backtracking

func subsets(nums []int) [][]int {
	return helper(nums, 0)
}

func helper(nums []int, start int) [][]int {
	res := [][]int{}
	res = append(res, []int{})
	if start == len(nums) {
		return res
	}

	for i := start; i < len(nums); i++ {
		tmp := helper(nums, i+1)
		for j := range tmp {
			t := append([]int{nums[i]}, tmp[j]...)
			res = append(res, t)
		}
	}
	return res
}

File: leetcode/BackTracking/leetcode-78/78.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> set = new ArrayList<>();
        res.add(new ArrayList<Integer>(set));

        if (nums == null || nums.length == 0) {
            return res;
        }

        for (int i = 0; i < nums.length; i++) {
            helper(nums, i, res, set);
        }
        return res;
    }

    public void helper(int[] nums, int start, List<List<Integer>> res, List<Integer> set) {
        if (start >= nums.length) {
            return;
        }

        set.add(nums[start]);
        res.add(new ArrayList<Integer>(set));
        for (int i = start + 1; i < nums.length; i++) {
            helper(nums, i, res, set);
        }
        set.remove(set.size() - 1);
    }
}